src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,190 @@
+        target.prev = newNode;
+        previousNode.next = newNode;
+        newNode.prev = previousNode;
+    }
+
+    private Node<T> getNode(int index) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("invalid index");
+        }
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode.data;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("invalid index");
+        }
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        T oldData = currentNode.data;
+        currentNode.data = value;
+        return oldData;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode;
+        if (index == 0) {
+            currentNode = head;
+            head = head.next;
+            currentNode.next.prev = null;
+            size--;
+            return currentNode.data;
+        }
+        currentNode = getNode(index);
+        if (currentNode.next == null) {
+            currentNode.prev.next = null;
+            size--;
+            return currentNode.data;
+        }
+        unlink(currentNode);
+        size--;
+        return currentNode.data;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (object == null ? currentNode.data == null : object.equals(currentNode.data)) {
+                if (size == 1) {
+                    head = new Node<>(null, tail, null);
+                    size--;
+                    return true;
+                }
+                if (currentNode == head) {
+                    head = head.next;
+                    head.prev = null;
+                    size--;
+                    return true;
+                }
+                if (currentNode.next == null) {
+                    currentNode.prev.next = null;
+                    size--;
+                    return true;
+                }
+                unlink(currentNode);
+                size--;
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
+    private void unlink(Node<T> node) {
```

Comment:
"size--" should be in "unlink" method.
Because "unlink" is always followed by "size--"